### PR TITLE
Finish refactor of UID/GID usage to a new struct

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
Finished the refactoring which was started in #33362: replace all remaining usages of pairs of `[]idtools.IDMap` with `idtools.IdentityMapping`.

**- How I did it**
To ensure that there would be no change in semantics with uninitialized fields, I replaced `[]IDMap` struct fields with `IdentityMapping` values and refactored all code to pass `IdentityMapping` structs by value. BuildKit makes heavy use of pass-by-reference `*IdentityMapping` values; I handled that as a special case in the builder-next adapters.

Changing the `IdentityMapping` associated functions from pointer receivers to value receivers is a non-breaking change as the pointer-receiver version of the functions would also panic when called on a nil pointer. The only notable difference would be in the stack trace: the nil dereference will now occur in the implicit dereference at the call site, rather than inside the implementation.

**- How to verify it**
CI

**- Description for the changelog**
N/A

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Starting do do a first review; w.r.t. the first commit; perhaps it's still possible to change (if it's only buildkit consuming it); I'll let @tonistiigi and @crazy-max comment on that;

> IdentityMapping values are already used extensively within BuildKit so
we cannot do that without also refactoring BuildKit in lockstep.

File: daemon/graphdriver/aufs/aufs.go
Comment (thaJeztah) [inline daemon/graphdriver/aufs/aufs.go]: Wondering why you're introducing the deprecated utility in these intermediate commits, before replacing them later, or was that just to make it easier to find back which locations had to be updated? https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/pkg/idtools/idtools.go#L119

File: pkg/idtools/idtools.go
Comment (thaJeztah) [inline pkg/idtools/idtools.go:177]: Slightly wondering if this would still be useful / convenient to have (to handle the nil cases), but I'd have to check again where / how it's all used (otherwise, good with me to clean it up 👍)

**Review by thaJeztah — APPROVED** [review]
nice! overall this looks quotes straightforward (and looks good to me); left one comment about the intermediate commits, but not too tied to that (just wondering if we could avoid the extra code churn)

File: pkg/idtools/idtools.go
Comment (corhere) [reply pkg/idtools/idtools.go:177]: Oh yeah, the nil cases... 🤦 I made a mistake. There's at least one case where an `*IdentityMapping` function argument is immediately dereferenced without any nil-pointer checks. That's no good; excuse me while I go revise this PR.

Comment (thaJeztah) [reply pkg/idtools/idtools.go:177]: Oh! Good that I commented then (probably shouldn't have "LGTM'd while on the road 🤦); no worries!

File: daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go
Comment (corhere) [inline daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go:188]: I noticed that `root == dirID` in this driver so I coalesced the two when I refactored it. But I noticed that other drivers follow a very similar pattern, except that their `root != dirID`: they set `dirID.UID` to `idtools.CurrentIdentity().UID`. Could this be a bug in the fuseoverlayfs driver?

Comment (thaJeztah) [reply daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go:188]: ^^ @AkihiroSuda

**Comment by thaJeztah** [comment]
@AkihiroSuda @tonistiigi PTAL 🙏

**Review by AkihiroSuda — APPROVED** [review]

File: daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go
Comment (AkihiroSuda) [reply daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go:188]: Yes, seems a bug

Comment (thaJeztah) [reply daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go:188]: Thanks for looking! Ok, given that this PR should not make the situation "worse"; let's fix that separately.

@corhere do you want to work on that? (Otherwise feel free to create a tracking issue for it)

**Review by thaJeztah — APPROVED** [review]
still LGTM, thanks!

**Pull request merged**
